File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/crownhunt/FirebaseCrownHuntStatsRepository.kt
package com.kungsbackacarcommunity.app.crownhunt

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.kungsbackacarcommunity.app.firebase.awaitOrThrow
import com.kungsbackacarcommunity.app.profile.FirebaseLiveProfileRepository
import com.kungsbackacarcommunity.app.profile.LiveProfileRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flow

/**
 * [CrownHuntStatsRepository] backed by rules-gated Firestore reads of the #710
 * aggregates:
 *
 *  - `crownHuntLeaderboardEntries` (scope `YYYY-MM`) — this season's ranked board,
 *    read `where scope == season orderBy points desc, crownsCollected desc limit N`
 *    (the deployed composite index `scope ASC, points DESC, crownsCollected DESC`),
 *    plus the viewer's own `alltime__{uid}` and `{season}__{uid}` counter docs.
 *  - `crownHuntUserStats/{uid}` — the viewer's own streak / rarity / seasonsWon,
 *    readable only by the owner.
 *
 * Display names are resolved from the members' public `users/{uid}` profiles via
 * the shared [LiveProfileRepository], best-effort: a missing profile falls back to
 * a short uid stub and never blocks the board. Nothing here writes — every one of
 * these collections is backend-trigger-owned.
 */
class FirebaseCrownHuntStatsRepository(
    private val firestore: FirebaseFirestore,
    private val liveProfiles: LiveProfileRepository,
    private val seasonIdProvider: () -> String = { CrownSeasonClock.currentSeasonId() },
) : CrownHuntStatsRepository {

    override fun observeStats(uid: String): Flow<CrownStatsUiState> =
        flow {
            emit(CrownStatsUiState.Loading)
            val seasonId = seasonIdProvider()

            // This season's ranked page. Shaped to the deployed composite index;
            // all conditions on the query, as the security rule expects.
            val boardSnap =
                firestore
                    .collection(LEADERBOARD)
                    .whereEqualTo(FIELD_SCOPE, seasonId)
                    .orderBy(FIELD_POINTS, Query.Direction.DESCENDING)
                    .orderBy(FIELD_CROWNS, Query.Direction.DESCENDING)
                    .limit(CrownHuntBoard.LEADERBOARD_TOP_N.toLong())
                    .get()
                    .awaitOrThrow { "crown leaderboard query failed without a cause" }
            val counters = boardSnap.documents.mapNotNull { it.toCounter() }

            // The viewer's own counters (may be absent — they have never
            // collected in that scope) and rich stats.
            val seasonEntry =
                firestore.collection(LEADERBOARD).document(entryId(seasonId, uid))
                    .get().awaitOrThrow { "crown season entry read failed" }
            val allTimeEntry =
                firestore.collection(LEADERBOARD)
                    .document(entryId(CrownSeasonClock.ALL_TIME_SCOPE, uid))
                    .get().awaitOrThrow { "crown all-time entry read failed" }
            val statsDoc =
                firestore.collection(USER_STATS).document(uid)
                    .get().awaitOrThrow { "crown user-stats read failed" }

            // Resolve names for the ranked page and the viewer.
            val uids = counters.map { it.uid }.toMutableSet().apply { add(uid) }
            val names =
                liveProfiles.loadProfiles(uids)
                    .mapNotNull { (id, profile) ->
                        profile.displayName?.let { id to it }
                    }
                    .toMap()

            val board = CrownHuntBoard.board(counters, uid, names, seasonId)
            val personal =
                CrownHuntBoard.personalStats(
                    allTime = allTimeEntry.toCounter(),
                    season = seasonEntry.toCounter(),
                    seasonRank = board.viewerRank,
                    rich = statsDoc.toUserStats(),
                )
            emit(CrownStatsUiState.Loaded(personal = personal, board = board))
        }.catch { emit(CrownStatsUiState.Error) }

    companion object {
        private const val LEADERBOARD = "crownHuntLeaderboardEntries"
        private const val USER_STATS = "crownHuntUserStats"
        private const val FIELD_SCOPE = "scope"
        private const val FIELD_POINTS = "points"
        private const val FIELD_CROWNS = "crownsCollected"

        /** Mirrors the backend `leaderboardEntryRef` id: `{scope}__{uid}`. */
        private fun entryId(scope: String, uid: String): String = "${scope}__$uid"

        /**
         * A stats repository, or null in a config-less/CI build (no Firebase) — so
         * the hub page simply shows its loading/empty affordance rather than
         * crashing, exactly as the crown map layer degrades.
         */
        fun createIfAvailable(context: Context): CrownHuntStatsRepository? {
            if (FirebaseApp.getApps(context).isEmpty()) return null
            return FirebaseCrownHuntStatsRepository(
                FirebaseFirestore.getInstance(),
                // Reuse the shared, cached live-profile repo so name resolution
                // shares the app's one read scope rather than opening a second.
                FirebaseLiveProfileRepository.sharedOrEmpty(context),
            )
        }
    }
}

/**
 * Maps a `crownHuntLeaderboardEntries/{scope}__{uid}` document to a counter, or
 * null when it does not exist or is missing its uid (a document that cannot be
 * ranked is dropped rather than shown as an anonymous zero row).
 */
private fun DocumentSnapshot.toCounter(): CrownLeaderboardCounter? {
    if (!exists()) return null
    val uid = getString("uid") ?: return null
    return CrownLeaderboardCounter(
        uid = uid,
        points = (get("points") as? Number)?.toInt() ?: 0,
        crownsCollected = (get("crownsCollected") as? Number)?.toInt() ?: 0,
    )
}

/** Maps a `crownHuntUserStats/{uid}` document to the hub's read subset. */
private fun DocumentSnapshot.toUserStats(): CrownUserStatsDoc? {
    if (!exists()) return null
    @Suppress("UNCHECKED_CAST")
    val rarityMap = get("byRarity") as? Map<String, Any?> ?: emptyMap()
    val byRarity =
        CrownRarity.entries.mapNotNull { rarity ->
            (rarityMap[rarity.wire] as? Number)?.toInt()?.takeIf { it > 0 }?.let { rarity to it }
        }.toMap()
    return CrownUserStatsDoc(
        byRarity = byRarity,
        // The streak fields are written under these keys by the ledger trigger.
        streakCurrent = (get("collectionStreakCurrent") as? Number)?.toInt() ?: 0,
        streakBest = (get("collectionStreakBest") as? Number)?.toInt() ?: 0,
        seasonsWon = (get("seasonsWon") as? Number)?.toInt() ?: 0,
        rarest = CrownRarity.fromWire(getString("rarestRarity")),
    )
}

File: apps/ios/KCC/CrownHunt/FirebaseCrownHuntStatsRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

/// ``CrownHuntStatsRepository`` backed by rules-gated Firestore reads of the
/// read-optimised aggregates — the iOS port of Android's
/// `FirebaseCrownHuntStatsRepository`.
///
///  - `crownHuntLeaderboardEntries` (scope `YYYY-MM`) — this season's ranked
///    board, read `where scope == season orderBy points desc, crownsCollected
///    desc limit N` (the deployed composite index), plus the viewer's own
///    `alltime__{uid}` and `{season}__{uid}` counter docs by id.
///  - `crownHuntUserStats/{uid}` — the viewer's own streak / rarity / seasonsWon,
///    readable only by the owner.
///
/// Display names are resolved from the members' public `users/{uid}` profiles,
/// best-effort: a missing profile falls back to a short uid stub and never
/// blocks the board. The read is one-shot per subscription (the aggregates move
/// only when someone collects a crown). Nothing here writes — every one of
/// these collections is backend-trigger-owned. Construction is guarded
/// (``createIfAvailable()`` returns nil without Firebase).
final class FirebaseCrownHuntStatsRepository: CrownHuntStatsRepository, @unchecked Sendable {
    private let firestore: Firestore
    private let seasonIdProvider: @Sendable () -> String

    private init(
        firestore: Firestore,
        seasonIdProvider: @escaping @Sendable () -> String
    ) {
        self.firestore = firestore
        self.seasonIdProvider = seasonIdProvider
    }

    func stats(uid: String) -> AsyncStream<CrownStatsSnapshot> {
        AsyncStream { continuation in
            let task = Task {
                do {
                    let snapshot = try await self.readStats(uid: uid)
                    continuation.yield(snapshot)
                } catch {
                    continuation.yield(.failed(code: CrownHuntFirestore.statusName(error)))
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func readStats(uid: String) async throws -> CrownStatsSnapshot {
        let seasonId = seasonIdProvider()

        // This season's ranked page. Shaped to the deployed composite index;
        // all conditions on the query, as the security rule expects.
        let boardSnap = try await firestore
            .collection(Self.leaderboard)
            .whereField(Self.scopeField, isEqualTo: seasonId)
            .order(by: Self.pointsField, descending: true)
            .order(by: Self.crownsField, descending: true)
            .limit(to: CrownHuntBoard.leaderboardTopN)
            .getDocuments()
        let counters = boardSnap.documents.compactMap(Self.counter(from:))

        // The viewer's own counters (may be absent) and rich stats — three
        // independent reads, fetched concurrently rather than sequentially.
        async let seasonEntryTask = firestore
            .collection(Self.leaderboard).document(Self.entryId(seasonId, uid)).getDocument()
        async let allTimeEntryTask = firestore
            .collection(Self.leaderboard)
            .document(Self.entryId(CrownSeasonClock.allTimeScope, uid))
            .getDocument()
        async let statsDocTask = firestore
            .collection(Self.userStats).document(uid).getDocument()
        let (seasonEntry, allTimeEntry, statsDoc) = try await (
            seasonEntryTask, allTimeEntryTask, statsDocTask
        )

        // Resolve names for the ranked page and the viewer, best-effort.
        var uids = Set(counters.map { $0.uid })
        uids.insert(uid)
        let names = await resolveNames(uids)

        let board = CrownHuntBoard.board(
            counters: counters, viewerUid: uid, names: names, seasonId: seasonId
        )
        let personal = CrownHuntBoard.personalStats(
            allTime: Self.counter(from: allTimeEntry),
            season: Self.counter(from: seasonEntry),
            seasonRank: board.viewerRank,
            rich: Self.userStats(from: statsDoc)
        )
        return .loaded(CrownStatsData(personal: personal, board: board))
    }

    /// Best-effort display-name resolution: reads each public `users/{uid}`
    /// profile by id, dropping any that error or lack a name. Bounded to the
    /// ranked page (<= LEADERBOARD_TOP_N) plus the viewer, and fetched
    /// concurrently (a `withTaskGroup` fan-out, not a sequential loop) so the
    /// stats read's latency is one round trip, not the sum of up to
    /// LEADERBOARD_TOP_N + 1 of them.
    ///
    /// Genuinely best-effort: a single profile read failing (deleted account,
    /// a transient permission hiccup) is swallowed here rather than thrown, so
    /// it degrades that one row to the uid stub instead of failing the whole
    /// stats read.
    private func resolveNames(_ uids: Set<String>) async -> [String: String] {
        await withTaskGroup(of: (uid: String, name: String?).self) { group in
            for uid in uids {
                group.addTask {
                    guard
                        let document = try? await self.firestore.collection(Self.users)
                            .document(uid).getDocument()
                    else { return (uid, nil) }
                    return (uid, document.get(Self.displayNameField) as? String)
                }
            }
            var names: [String: String] = [:]
            for await (uid, name) in group {
                if let name { names[uid] = name }
            }
            return names
        }
    }

    // MARK: - Mapping

    /// Maps a `crownHuntLeaderboardEntries/{scope}__{uid}` document to a counter,
    /// or nil when it does not exist or is missing its uid.
    static func counter(from document: DocumentSnapshot) -> CrownLeaderboardCounter? {
        guard document.exists, let uid = document.get("uid") as? String else { return nil }
        return CrownLeaderboardCounter(
            uid: uid,
            points: (document.get(pointsField) as? NSNumber)?.intValue ?? 0,
            crownsCollected: (document.get(crownsField) as? NSNumber)?.intValue ?? 0
        )
    }

    /// Maps a `crownHuntUserStats/{uid}` document to the hub's read subset.
    static func userStats(from document: DocumentSnapshot) -> CrownUserStatsDoc? {
        guard document.exists else { return nil }
        let rarityMap = document.get("byRarity") as? [String: Any] ?? [:]
        var byRarity: [CrownRarity: Int] = [:]
        for rarity in CrownRarity.allCases {
            if let count = (rarityMap[rarity.wire] as? NSNumber)?.intValue, count > 0 {
                byRarity[rarity] = count
            }
        }
        return CrownUserStatsDoc(
            byRarity: byRarity,
            // The streak fields are written under these keys by the ledger trigger.
            streakCurrent: (document.get("collectionStreakCurrent") as? NSNumber)?.intValue ?? 0,
            streakBest: (document.get("collectionStreakBest") as? NSNumber)?.intValue ?? 0,
            seasonsWon: (document.get("seasonsWon") as? NSNumber)?.intValue ?? 0,
            rarest: CrownRarity.fromWire(document.get("rarestRarity") as? String)
        )
    }

    /// Mirrors the backend `leaderboardEntryRef` id: `{scope}__{uid}`.
    private static func entryId(_ scope: String, _ uid: String) -> String { "\(scope)__\(uid)" }

    // MARK: - Factory

    private static let leaderboard = "crownHuntLeaderboardEntries"
    private static let userStats = "crownHuntUserStats"
    private static let users = "users"
    private static let scopeField = "scope"
    private static let pointsField = "points"
    private static let crownsField = "crownsCollected"
    private static let displayNameField = "displayName"

    private static let cachedLock = NSLock()
    nonisolated(unsafe) private static var cached: FirebaseCrownHuntStatsRepository?

    /// A stats repository, or nil in a config-less/CI build (no Firebase) — so
    /// the hub shows its unavailable affordance rather than crashing, exactly as
    /// every other Firebase-backed surface degrades. Honors the
    /// `FIREBASE_FIRESTORE_EMULATOR_HOST` seam.
    static func createIfAvailable(
        seasonIdProvider: @escaping @Sendable () -> String = { CrownSeasonClock.currentSeasonId() }
    ) -> CrownHuntStatsRepository? {
        guard FirebaseApp.app() != nil else { return nil }
        cachedLock.lock()
        defer { cachedLock.unlock() }
        if let cached { return cached }
        let firestore = CrownHuntFirebase.firestore()
        let repository = FirebaseCrownHuntStatsRepository(
            firestore: firestore, seasonIdProvider: seasonIdProvider
        )
        cached = repository
        return repository
    }
}
